cpu/rpx0xx/include/pio/pio.h

```diff
@@ -0,0 +1,1166 @@
+    PIO_INST_MOV_DST_ISR                = 6,    /**< To input shift register */
+    PIO_INST_MOV_DST_OSR                = 7     /**< To output shift register */
+} pio_inst_mov_dst_t;
+/**
+ * @brief MOV operation
+ */
+typedef enum {
+    PIO_INST_MOV_OP_NONE                = 0,    /**< No operation */
+    PIO_INST_MOV_OP_INVERT              = 1,    /**< Bitwise complement */
+    PIO_INST_MOV_OP_REVERSE             = 2     /**< Bitwise reverse */
+} pio_inst_mov_op_t;
+/**
+ * @brief   MOV source
+ */
+typedef enum {
+    PIO_INST_MOV_SRC_PINS               = 0,    /**< From input pins */
+    PIO_INST_MOV_SRC_X                  = 1,    /**< From scratch X */
+    PIO_INST_MOV_SRC_Y                  = 2,    /**< From scratch Y */
+    PIO_INST_MOV_SRC_NULL               = 3,    /**< Fill with zeros */
+    PIO_INST_MOV_SRC_STATUS             = 5,    /**< From status */
+    PIO_INST_MOV_SRC_ISR                = 6,    /**< From input shift register */
+    PIO_INST_MOV_SRC_OSR                = 7     /**< From output shift register */
+} pio_inst_mov_src_t;
+/**
+ * @brief   IRQ opcode
+ */
+#define PIO_INST_IRQ                    (6u << 13)
+/**
+ * @brief   IRQ opcode mask
+ */
+#define PIO_INST_IRQ_MASK               ((7u << 13) | (1u << 7))
+/**
+ * @brief   IRQ clear flag shift position
+ */
+#define PIO_INST_IRQ_CLR_SHIFT          6
+/**
+ * @brief   IRQ clear flag mask
+ */
+#define PIO_INST_IRQ_CLR_MASK           (1u << PIO_INST_IRQ_CLR_SHIFT)
+/**
+ * @brief   IRQ wait until clreared flag shift position
+ */
+#define PIO_INST_IRQ_WAIT_SHIFT         5
+/**
+ * @brief   IRQ wait until cleared flag mask
+ */
+#define PIO_INST_IRQ_WAIT_MASK          (1u << PIO_INST_IRQ_WAIT_SHIFT)
+/**
+ * @brief   IRQ index mask
+ */
+#define PIO_INST_IRQ_INDEX_MASK         (31u)
+/**
+ * @brief   SET opcode
+ */
+#define PIO_INST_SET                    (7u << 13)
+/**
+ * @brief   SET opcode mask
+ */
+#define PIO_INST_SET_MASK               (7u << 13)
+/**
+ * @brief   SET destination shift position
+ */
+#define PIO_INST_SET_DESTINATION_SHIFT  5
+/**
+ * @brief   SET destination mask
+ */
+#define PIO_INST_SET_DESTINATION_MASK   (7u << PIO_INST_SET_DESTINATION_SHIFT)
+/**
+ * @brief   SET data mask
+ */
+#define PIO_INST_SET_DATA_MASK          (31u)
+/**
+ * @brief   SET destinations
+ */
+typedef enum {
+    PIO_INST_SET_DST_PINS               = 0,    /**< To set pins */
+    PIO_INST_SET_DST_X                  = 1,    /**< To scratch X */
+    PIO_INST_SET_DST_Y                  = 2,    /**< To scratch Y */
+    PIO_INST_SET_DST_PINDIRS            = 4     /**< To set pin directions */
+} pio_inst_set_dst_t;
+
+/**
+ * @brief   Construct a JMP instruction
+ *
+ * @param[in]       condition       Jump if this condition is true
+ * @param[in]       address         Address in instruction memory to jump to
+ *
+ * @return  JMP instruction
+ */
+static inline pio_instr_t pio_inst_jmp(pio_inst_jmp_cond_t condition,
```

Comment:
I think there should be a bit explaining how those manually injected instructions are supposed to be used.

Will they interrupt the PIO program flow at any point or is it only possible to insert them when the PIO is in a certain state?

Reply:
Explanation added to `pio_sm_exec()`